text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+3. the `upgrade-bootstrapper` spawns a child process, running the `casper-node` binary in validator mode, passing it the
+   path to its `config.toml` as an arg
+4. while they are running, before the network reaches the upgrade point, the next version of the `casper-node` gets
+   installed, but not run.  The binary is always named `casper-node-next` and its config named `config-next.toml`.  Each
+   gets installed to the same location as the current equivalent files.  The location of the `config.toml` can be
+   retrieved from the known location used by the `upgrade-bootstrapper` at step 2.  An upgrade file for chainspec
+   modifications should also be installed and its path included in the new `config-next.toml`
+5. when the current upgrade activation point is reached and the node has finished all its work in that era, it exits
+   normally with an exit code 0
+6. the `upgrade-bootstrapper` renames
+     * `casper-node-next` to `casper-node` (replaces the old binary)
+     * `config.toml` to `config-old.toml`
+     * `config-next.toml` to `config.toml`
+7. the `upgrade-bootstrapper` runs the new `casper-node` in migrate-data mode (see below for further details), passing
+   it the path to `config.toml` and `config-old.toml` as args.  The node performs any migration of stored data and
+   config options required, then exits normally with an exit code 0
+8. repeat from step 3
+
+Should the child process stop with a non-zero exit code at any stage, the `upgrade-bootstrapper` should itself panic.
+It is up to the user to restart the `upgrade-bootstrapper` manually.
+
+It is assumed that the `casper-node` binary will live at a known location and that installing an upgraded binary will
+place the binary and its config file in a known location.  If we wish to be more flexible about this (particularly the
+location of the `casper-node` binary) we can handle these locations in a similar way to the location of the
+`casper-node`'s `config.toml`.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to existing `casper-node` (running in validator mode)
+
+We require the following changes to `casper-node`:
+
+* running in validator mode, when we reach the upgrade activation point, exit gracefully
+* running in validator mode, if the node starts, joins and finds that it has already reached the upgrade activation
+  point, exit gracefully
+* upgrade activation points should be defined in terms of era IDs, not block heights
+* config.toml updated to take the path to the chainspec upgrade file, similar to the `node.chainspec_config_path`
+  option.  Alternatively, just remove this option, and require the chainspec, its upgrade file, accounts.csv and the
+  config.toml file all to live in the same folder
+
+### Data migration
+
+We require to add a new `migrate-data` subcommand to `casper-node`.  We could provide a separate binary to be used
```

Comment:
An alternative would be to store the version number in the storage directory, so the new process knows what version the data belongs to and can determine whether it needs to perform the migration. (Not sure if that's better, but it would simplify the interface between the bootstrapper and the node.)

Reply:
Yes, fair point.

My thinking was to try and keep the "normal" operation of node as far removed from the migration operation as possible, to the point where I considered a separate binary just for this.  Mainly because this should be the only place in our codebase where we need to handle two different versions of data or config types at the same time.  I'd like that sort of stuff to be kept away from the core of our current node codebase as practicable.  I was envisaging this not even touching our current storage component.

In terms of simplifying the interface, yes and no :)  The bootstrapper would only have to invoke the node once rather than twice.  However, to migrate the config data, it would need to take the path to the old config file too.  For the very first run, we'd either have to make the old path optional and not pass anything (slightly dangerous since it should always be required thereafter) or else we could just pass the same path twice, implying no migration is required (but that's a weird interface).